ibis/backends/flink/tests/test_join.py

```diff
@@ -0,0 +1,165 @@
+    df = df[list(schema.names)]
+    df = df.head()
+
+    # Create `table_left`
+    table_left_tempdir = tempfile.TemporaryDirectory()
+    table_left = con.create_table(
+        "table_left",
+        schema=schema,
+        tbl_properties=tempdir_sink_configs(table_left_tempdir.name),
+        watermark=ibis.watermark(
+            time_col="timestamp_col", allowed_delay=ibis.interval(seconds=15)
+        ),
+    )
+    con.insert(
+        "table_left",
+        obj=df,
+        schema=schema,
+    ).wait()
+
+    # Create `table_left`
+    table_right_tempdir = tempfile.TemporaryDirectory()
+    table_right = con.create_table(
+        "table_right",
+        schema=schema,
+        tbl_properties=tempdir_sink_configs(table_right_tempdir.name),
+        watermark=ibis.watermark(
+            time_col="timestamp_col", allowed_delay=ibis.interval(seconds=15)
+        ),
+        primary_key="id",
+    )
+    con.insert(
+        "table_right",
+        obj=df,
+        schema=schema,
+    ).wait()
+
+    # Join `table_left` and `table_right`
+    expr = table_left.asof_join(
+        table_right,
+        predicates=[
+            table_left["id"] == table_right["id"],
+            table_left["timestamp_col"] >= table_right["timestamp_col"],
+        ],
+    )
+    sql = con.compile(expr)
+
+    expected_sql = """SELECT t0.`id`, t0.`bool_col`, t0.`smallint_col`, t0.`int_col`,
+       t0.`timestamp_col`, CAST(t1.`id` AS int) AS `id_right`,
+       t1.`bool_col` AS `bool_col_right`,
+       t1.`smallint_col` AS `smallint_col_right`,
+       t1.`int_col` AS `int_col_right`,
+       t1.`timestamp_col` AS `timestamp_col_right`
+FROM table_left AS t0
+  JOIN table_right FOR SYSTEM_TIME AS OF t0.`timestamp_col` AS t1
+    ON (t0.`id` = t1.`id`)"""
+    assert sql == expected_sql
+
+    join_df = expr.to_pandas().sort_values("id")
+
+    table_left_tempdir.cleanup()
+    table_right_tempdir.cleanup()
+
+    expected_df_to_str = (
+        "  id  bool_col  smallint_col  int_col           timestamp_col  id_right  bool_col_right  smallint_col_right  int_col_right     timestamp_col_right\n"
+        "6690      True             0        0 2010-11-01 00:00:00.000      6690            True                   0              0 2010-11-01 00:00:00.000\n"
+        "6691     False             1        1 2010-11-01 00:01:00.000      6691           False                   1              1 2010-11-01 00:01:00.000\n"
+        "6692      True             2        2 2010-11-01 00:02:00.100      6692            True                   2              2 2010-11-01 00:02:00.100\n"
+        "6693     False             3        3 2010-11-01 00:03:00.300      6693           False                   3              3 2010-11-01 00:03:00.300\n"
+        "6694      True             4        4 2010-11-01 00:04:00.600      6694            True                   4              4 2010-11-01 00:04:00.600"
+    )
+    assert join_df.to_string(index=False) == expected_df_to_str
```

Comment:
also wondering if it's prefered to use `tm.assert_frame_equal()` here instead? i haven't seen the comparison this way in ibis unit tests

Reply:
Good point. I actually tried to validate with something like the following:
```python
    expected = pd.merge_asof(
        df, df,
        on="timestamp_col",
        by="id",
    )

    tm.assert_frame_equal(result, expected)
```
This however does not work as intended since the number of columns in `expected` will be less than that in `result` due to the lack of proper deduplication of columns for temporal join, [per this this thread](https://github.com/ibis-project/ibis/pull/7921#discussion_r1449132806).